implement refactor/redesign stages of `flox activate` (#1299)
flox/flox: cli/flox/doc/manifest.toml.md

SERVER_PORT = "3000"
```

## `[hook]`

> The `[hook]` section of the manifest allows you to specify scripts to be invoked by a

Author: I'm a huge fan of sembr.org - was trying to remain consistent with existing lines in this file but happy to buck that trend and stick with semantic line breaks.  👍

**bash** shell immediately upon activating an environment for the *first time only*.
Subsequent activations, such as the "in-place" activations performed by way of
"in-place" invocations performed in subshell "dotfiles" will *NOT* re-run these scripts.

Hook scripts inherit environment variables set in the `[vars]` section,
and variables set by the script will in turn be inherited by the `[profile]`

> scripts described below. Any stdout generated by the scripts in this section

Author: included in patch below - resolving conversation

> Hook scripts inherit environment variables set in the `[vars]` section,
> and variables set by the script will in turn be inherited by the `[profile]`
> scripts described below. Any stdout generated by the scripts in this section
> will be redirected to stderr to avoid it being interpreted as code for
> "in-place" activations.

Author: ```suggestion                                                                                                                                                                                                                                                                   
Hook scripts inherit environment variables set in the `[vars]` section,                                                                                                                                                                                                         
and variables set by the script will in turn be inherited by                                                                                                                                                                                                                    
the `[profile]` scripts described below.                                                                                                                                                                                                                                        
Any output written to `stdout` in a hook script is redirected to `stderr`                                                                                                                                                                                                       
to avoid it being mixed with the output of profile section scripts                                                                                                                                                                                                              
that must be able to write to `stdout` for "in-place" activations.                                                                                                                                                                                                              
```

The `on-activate` script is currently the only supported hook,
but we expect to add additional ones in future.

### `on-activate`

> The `on-activate` script is useful for performing bash syntax

Author: Thanks ... incorporated into patch below.

> initialization when first activing an environment, and for which

Author: done in daa59c46c

> you want the effects to be available to all supported shells,

Author: Was that a typo? Not sure that makes sense?

> you want the effects to be available to all supported shells,

Author: Section rewritten ... resolving conversation.

> e.g. spawning a background process, dynamically setting environment

Author: a processes?

> e.g. spawning a background process, dynamically setting environment

Author: Got it ... changed to plural.  👍

variables, or creating files and directories.
Environment variables declared in the `[vars]` section
can be used in this script.

```toml
[hook]
on-activate = """
    mkdir -p data_dir

>     # Start daemons which set env vars via stdout, e.g. a familiar example:

Author: Discussed above; will let Ross find a new/better example.

    eval $(ssh-agent)
"""
```

### `script` - DEPRECATED

> This `script` option was previously another name for the `on-activate` script,

Author: Yeah Yannik pointed this out as well; confusing history.

Thanks for the suggestion - patch below.

> This `script` option was previously another name for the `on-activate` script,
> and we renamed it to avoid ambiguity and more clearly delineate it from the
> scripts described in the `[profile]` section below.
> It will be removed in a later release.

Author: ```suggestion
This field was deprecated in favor of the `profile` section.
It will be removed in a later release.
```

## `[profile]`

Scripts defined in the `[profile]` section are sourced by *your shell*
immediately following any `[hook]` invocations, and consequently inherit environment
variables both set in those scripts and from the `[vars]` section above.
The `profile.common` script is sourced by all shells
and must therefore be compatible with all shells.
The `profile.bash` and `profile.zsh` scripts will only be sourced
by the corresponding shell after sourcing the `profile.common` script.

These scripts are useful for performing shell-specific customizations
such as setting aliases or configuring the prompt.
Scripts in this section are sourced with *every* activation,
including those performed by way of "in-place" activations
performed from a `.bashrc` or similar "dotfile".

```toml
[profile]
common = """
    fortune | lolcat
"""
bash = """

>     set -o vi

Author: @rossturk is still thinking about the best examples for the template, and I'll remember to copy that here as well once it has been finalised.

SERVER_PORT = "3000"
```

## `[hook]`

Scripts in the `[hook]` section
are useful for performing initialization
in a predictable bash shell environment.

### `on-activate`

The `on-activate` script is invoked only once upon activation
and can be useful for spawning processes,
dynamically setting environment variables,
and creating files and directories
to be used by the subsequent profile scripts, commands and shells.

If Flox detects that the environment has not yet been activated,
then it will source this script from a **bash** shell.

> This prevents it from being sourced in subsequent "in place" activations,

Reviewer: blocking: needs some explanation of how environment variables are handled from the hook. How flox activate (with hook that sets VAR) && export VAR=something-else && eval "$(flox activate)" will reset VAR

Author: We've already commented that this user-facing text falls into the product realm for maintenance over time, and the previous comments were that it was too long, so in Graham's absence Ross and I met to boil it down to the most succinct description. Let's set a deadline for you to meet with Ross and add whatever you think is necessary here as my input has already been incorporated.

> This prevents it from being sourced in subsequent "in place" activations,

Author: Matthew to discuss with Ross

such as those performed with `eval "$(flox activate)"`.

Hook scripts inherit environment variables set in the `[vars]` section,
and variables set here will in turn be inherited by
the `[profile]` scripts described below.

Any output written to `stdout` in a hook script is redirected to `stderr`
to avoid it being mixed with the output of profile section scripts
that write to `stdout` for "in-place" activations.

> The `on-activate` script is currently the only supported hook.

Author: Adding to list for you and Ross to discuss.

The `profile.common` script is sourced
for every shell
and special care should be taken
to ensure compatibility with all shells.
The `profile.bash` and `profile.zsh` scripts
are then sourced
by the corresponding shell.

These scripts are useful for performing shell-specific customizations
such as setting aliases or configuring the prompt.
Scripts in this section are sourced with *every* activation,
including those performed by way of "in-place" activations
performed from a `.bashrc` or similar "dotfile".

```toml
[profile]
common = """
    fortune
"""
bash = """
    source $venv_dir/bin/activate
    alias foo="echo bar"
    set -o vi
"""
zsh = """
    source $venv_dir/bin/activate
    alias foo="echo bar"
    bindkey -v
"""
```

The `profile.common` script
is intended for setup that will be sourced for every shell,
and it is your responsibility
to make sure that commands are compatible with all shells
in which the environment is expected to be activated.
The `profile.bash` and `profile.zsh` scripts
are sourced *after* `profile.common`,
and are only sourced by the corresponding shell.
The shell-specific profile scripts
are intended for functions, aliases, variables, etc.

> that could be specific to a user's shell.

Author: Adding to list for you and Ross to discuss.
